lib/_http_outgoing.js

```diff
@@ -878,9 +878,10 @@ OutgoingMessage.prototype.end = function end(chunk, encoding, callback) {
 
   if (this._hasBody && this.chunkedEncoding) {
     this._send('0\r\n' + this._trailer + '\r\n', 'latin1', finish);
-  } else {
-    // Force a flush, HACK.
+  } else if (!this._headerSent || this.outputSize || chunk) {
     this._send('', 'latin1', finish);
+  } else {
+    process.nextTick(finish);
```

Comment:
The problem I see here is that the `'finish'` event might be emitted before buffered data is flushed and similarly the `response.end()` callback might be called before previous `response.write()` callbacks.

Reply:
I understand, but I *think* at this point there should not buffered data and a consecutive call to `end()` should not reach here, or am I missing smthg? Thanks

Reply:
What do you think about adding a check for `socket.writableLength` to the above `else if` branch?

```js
else if (!this._headerSent || this.outputSize || chunk || this.socket?.writableLength) {
  this._send('', 'latin1', finish);
}
```

Reply:
That sounds good. We might even use [`OutgoingMessage.writableLength`](https://github.com/nodejs/node/blob/master/lib/_http_outgoing.js#L161) directly. Thanks!

Reply:
@lpinca pushed the changes

Reply:
> I *think* at this point there should not buffered data and a consecutive call to `end()` should not reach here

I think you are right because the transfer encoding would be chunked but I'm not sure if there is are cases where that path can be taken anyway. For example in https://github.com/nodejs/node/issues/41062 the user specifies the `Content-Length` header so the chunked encoding is not used.